### ISSUE TITLE
Full support for UK CGT: Average cost basis and 30 day rule

### ISSUE THREAD

**Issue by dankrad** [rotki/rotki#3941]
UK tax rules use the average cost base, as specified in the issue #1793. In addition, there is also the so called bed and breakfast rule, also called 30 day rule. If an asset is sold, and then the same asset re-bought within 30 days, then the price at which it was re-bought is the cost basis.

The rule can be found here: https://www.gov.uk/hmrc-internal-manuals/capital-gains-manual/cg13370

As @mcplums mentioned in #1783, it is probably best to let a professional look over the implementation and check the details.

**Comment by LefterisJP** [comment]
Hey @dankrad since avg cost basis is coming next release I am changing this issue to be specifically for the bed and breakfast rule.

> it is probably best to let a professional look over the implementation and check the details.

We give no guarantees of correctness for anything. rotki is an opensource transparent tool that aims to be a customizable solution for as many jurisdiction as its users and contributors can implement. If a bug is found in the implementation we will fix it. If someone wants to audit an implementation or even better yet help implement it that would be amazing.

**Comment by aspiers** [comment]
Attempting some clarification here.  In addition to long-term average cost basis, there are _two_ extra rules which need implementing, rather than just one:

1. Disposal and reacquisition within 30 days
2. Disposal and reacquisition on the same day

If you follow [the link given in the original description above](https://www.gov.uk/hmrc-internal-manuals/capital-gains-manual/cg13370), you will see https://www.gov.uk/hmrc-internal-manuals/capital-gains-manual/cg51560 which clearly distinguishes between these and links to further related resources.

For a more human-friendly description, see https://koinly.io/guides/hmrc-cryptocurrency-tax-guide/#uk-cost-basis-method

**Comment by aspiers** [comment]
[As suggested a long time ago on Discord](https://discord.com/channels/657906918408585217/797933632609189888/911382752094486568), one approach to this might be to integrate Rotki's backend https://github.com/BittyTax/BittyTax, which just so happens to be Python and AGPL.

**Comment by aspiers** [comment]
If this was implemented as a PR, I would definitely compare the results for my accounts with what Koinly tells me and help to fix any discrepancies.

**Comment by LefterisJP** [comment]
Integration with such a different tool would not work.Reuing code or studying how it's supposed to work most probaby yes

**Comment by aspiers** [comment]
Yeah, it's unfortunate that BittyTax reimplemented a lot of what Rotki already does.

As [I previously noted](https://discord.com/channels/657906918408585217/797933632609189888/911388563491405916) (funnily, almost exactly a year ago!):

![image](https://user-images.githubusercontent.com/100738/206916103-27b06138-6132-4784-9bdb-7dbefa063817.png)

I can totally understand how it might not make since to build a straight integration between the two tools, given the huge overlap in functionality but significantly different approach to UI.

However it would be really nice if we could establish some kind of friendly collaborative agreement between the projects.  For example, at a glance it looks like there are bits of BittyTax's share pooling logic such as [`bittytax/tax.py`](https://github.com/BittyTax/BittyTax/blob/master/bittytax/tax.py) which could be really useful to Rotki, and maybe are already _almost_ reusable by Rotki.  In that case, maybe some kind of agreement could potentially be reached on how to:

1. make those bits 100% reusable in Rotki, and
2. ensure that they _stay_ reusable, with patches to those bits of code being able to flow in both directions between both projects, for an on-going win/win situation.

**Comment by dankrad** [comment]
> Attempting some clarification here. In addition to long-term average cost basis, there are _two_ extra rules which need implementing, rather than just one:
> 
>     1. Disposal and reacquisition within 30 days
> 
>     2. Disposal and reacquisition on the same day
> 
> 

Yes indeed, this seems to be correct.

**Comment by Arachnid** [comment]
Would love to see this implemented, it's really important for UK capital gains calculations.
